### PR TITLE
reduce allocations / GC related to articulated hands

### PR DESCRIPTION
**Pull request by davidkline-ms**
This change creates two static instances of WindowsMixedRealityHandProvider; one each for the left and right hands. This is done to reduce the number of GCs that occur as a result of hands coming in and out of the gesture frame.

Also changed was some initialization, improving the handling of the model and neutral poses changing during runtime.

Opening as a draft to get feedback while I test the code.

### CODE REVIEW

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityArticulatedHandDefinition.cs [hunk 1]
```diff
@@ -123,14 +126,23 @@ public void UpdateHandMesh(SpatialInteractionSourceState sourceState)
...
                 {
-                    if (handMeshTriangleIndices == null)
+                    if ((handMeshTriangleIndices == null) ||
+                        (handMeshTriangleIndices.Length != handMeshObserver.TriangleIndexCount))
                     {
                         handMeshTriangleIndices = new ushort[handMeshObserver.TriangleIndexCount];
                         handMeshTriangleIndicesUnity = new int[handMeshObserver.TriangleIndexCount];
+                    }
+
+                    if (!HandMeshModelId.HasValue || (HandMeshModelId.Value != handMeshObserver.ModelId))
```
Comment (keveleigh) [inline Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityArticulatedHandDefinition.cs:136]: Can this be simplified to
```suggestion
                    if (HandMeshModelId != handMeshObserver.ModelId)
```

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityArticulatedHandDefinition.cs [hunk 2]
```diff
@@ -123,14 +126,23 @@ public void UpdateHandMesh(SpatialInteractionSourceState sourceState)
...
+                    {
                         handMeshObserver.GetTriangleIndices(handMeshTriangleIndices);
+                        HandMeshModelId = handMeshObserver.ModelId;
 
                         Array.Copy(handMeshTriangleIndices, handMeshTriangleIndicesUnity, (int)handMeshObserver.TriangleIndexCount);
+                    }
 
+                    if (!NeutralPoseVersion.HasValue || (NeutralPoseVersion.Value != handMeshObserver.NeutralPoseVersion))
```
Comment (keveleigh) [inline Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityArticulatedHandDefinition.cs:144]: Also this
```suggestion
                    if (NeutralPoseVersion != handMeshObserver.NeutralPoseVersion)
```

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityHandMeshProvider.cs [hunk 3]
```diff
@@ -20,11 +20,33 @@ namespace Microsoft.MixedReality.Toolkit.WindowsMixedReality
...
+
+        [Obsolete("WindowsMixedRealityHandMeshProvider(IMixedRealityController) is obsolete. Please use either the static LeftHand or RightHand members and call SetController()")]
+        public WindowsMixedRealityHandMeshProvider(IMixedRealityController controller) => this.Controller = controller;
 
-        private readonly IMixedRealityController controller;
-        private IMixedRealityInputSource InputSource => controller.InputSource;
-        private Handedness Handedness => controller.ControllerHandedness;
+        private IMixedRealityController Controller = null;
+        private IMixedRealityInputSource InputSource => Controller.InputSource;
+        private Handedness Handedness;
```
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityHandMeshProvider.cs:37]: haha! i capitalized Controller, but Handedness was already that way 😄

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityArticulatedHandDefinition.cs (hunk 1 above)
Comment (davidkline-ms) [reply Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityArticulatedHandDefinition.cs:136]: Not if we use nullable ints.

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityArticulatedHandDefinition.cs (hunk 2 above)
Comment (davidkline-ms) [reply Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityArticulatedHandDefinition.cs:144]: see above

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityHandMeshProvider.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (keveleigh) [reply Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityHandMeshProvider.cs:37]: `Handedness` was a property before :)

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityArticulatedHandDefinition.cs (hunk 1 above)
Comment (keveleigh) [reply Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityArticulatedHandDefinition.cs:136]: but `null` should not equal any int, right? I thought C# was smart about checking nullable types against non-nullable ones :(

Comment (davidkline-ms) [reply Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityArticulatedHandDefinition.cs]: i may be thinking "old" c# 😄

**Review by davidkline-ms — COMMENTED** [review]
these changes look great @keveleigh!

**Review by keveleigh — APPROVED** [review]

**Pull request merged**